packages/syft/src/syft/service/code_history/code_history_service.py

```diff
@@ -0,0 +1,261 @@
+
+    @service_method(path="code_history.delete_by_id", name="delete_by_id")
+    def delete(self, context: AuthedServiceContext, uid: UID):
+        result = self.stash.delete_by_uid(context.credentials, uid)
+        if result.is_ok():
+            return result.ok()
+        else:
+            return SyftError(message=result.err())
+
+    def fetch_history_for_user(
+        self, context: AuthedServiceContext, user_verify_key
+    ) -> CodeHistoryDict:
+        result = self.stash.get_by_verify_key(
+            credentials=context.credentials, user_verify_key=user_verify_key
+        )
+        user_code_service = context.node.get_service("usercodeservice")
+
+        def get_code(uid):
+            return user_code_service.get_by_uid(context=context, uid=uid)
+
+        if result.is_ok():
+            code_histories = result.ok()
+            code_versions_dict = {}
+
+            for code_history in code_histories:
+                user_code_list = []
+                for uid in code_history.user_code_history:
+                    user_code_list.append(get_code(uid))
+                code_versions = CodeVersions(
+                    user_code_history=user_code_list,
+                    service_func_name=code_history.service_func_name,
+                    comment_history=code_history.comment_history,
+                )
+                code_versions_dict[code_history.service_func_name] = code_versions
+            return CodeHistoryDict(code_versions=code_versions_dict)
+        else:
+            return SyftError(message=result.err())
+
+    @service_method(
+        path="code_history.get_history",
+        name="get_history",
+        roles=DATA_SCIENTIST_ROLE_LEVEL,
+    )
+    def get_histories_for_current_user(self, context: AuthedServiceContext):
+        return self.fetch_history_for_user(
+            context=context, user_verify_key=context.credentials
+        )
+
+    @service_method(
+        path="code_history.get_history_for_user",
+        name="get_history_for_user",
+        roles=DATA_OWNER_ROLE_LEVEL,
+    )
+    def get_history_for_user(self, context: AuthedServiceContext, email: str):
+        user_service = context.node.get_service("userservice")
+        result = user_service.stash.get_by_email(
+            credentials=context.credentials, email=email
+        )
+        if result.is_ok():
+            user = result.ok()
+            return self.fetch_history_for_user(
+                context=context, user_verify_key=user.verify_key
+            )
+        return SyftError(message=result.err())
+
+    @service_method(
+        path="code_history.get_histories",
+        name="get_histories",
+        roles=DATA_OWNER_ROLE_LEVEL,
+    )
+    def get_histories_group_by_user(self, context: AuthedServiceContext):
+        result = self.stash.get_all(credentials=context.credentials)
+        if result.is_err():
+            return SyftError(message=result.err())
+        code_histories = result.ok()
```

Comment:
lets add a type, makes it much easier to read the code

Reply:
for code_histories?

Reply:
yes